# Part I - Loan Data from Prosper Exploration
## by Agustin Barto

In [1]:
from pathlib import Path
from urllib import request
from zipfile import ZipFile

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sb
from IPython.display import display, Markdown, HTML

%matplotlib inline

## Table of Contents

* [Introduction](#introduction)
* [Preliminary Wrangling](#preliminary-wrangling)
    * [Data Dictionary](#data-dictionary)

## Introduction<a class="anchor" id="introduction"></a>

I've chosen the "Loan Data from Prosper" dataset as it strikes a good balance between size, complexity and richness. It has a lot of variables to chose from, both numerical and categorical and enough samples to provided meaningful results.

As we'll show in the following section, the data set contains quite a lot of columns to chose from. During the wrangling process we'll briefly explore the dataset to decide which columns are going to be the focus of the analysis. Whenever needed, the column definition will be expanded with external sources.

## Preliminary Wrangling<a class="anchor" id="preliminary-wrangling"></a>

### Data dictionary<a class="anchor" id="data-dictionary"></a>

 The following table (converted to CSV from the original [Google sheet](https://docs.google.com/spreadsheets/d/1gDyi_L4UvIrLTEC6Wri5nbaMmkGmLQBk-Yx3z0XDEtI/edit#gid=0)) contains a brief description of each column:

In [2]:
variable_definitions_df = pd.read_csv(
    "https://docs.google.com/spreadsheets/d/1gDyi_L4UvIrLTEC6Wri5nbaMmkGmLQBk-Yx3z0XDEtI/export?format=csv",
)

In [3]:
styler = variable_definitions_df\
    .style\
    .set_properties(
        **{'text-align': 'left'}
    )\
    .hide(axis="index")
display(HTML(styler.to_html()))

### Downloading the dataset

We assume the dataset has been included with the submission, but in case it had to be removed due to size constraints, the following cell checks if the data is available and downloads it if it is not.

In [4]:
prosper_load_data_csv_zip = Path("./prosperLoanData.csv.zip")

In [5]:
def download_compress_dataset():
    try:
        with request.urlopen("https://s3.amazonaws.com/udacity-hosted-downloads/ud651/prosperLoanData.csv") as f:
            with ZipFile(prosper_load_data_csv_zip, "w") as zf:
                with zf.open("prosperLoanData.csv", "w") as g:
                    g.write(f.read())
    except Exception as e:
        display(Markdown(f"> <span style='color: red;'>**Exception raised retrieving data set:**</span> {e}"))
        prosper_load_data_csv_zip.unlink()

In [6]:
if not prosper_load_data_csv_zip.exists():
    download_compress_dataset()

### Loading the dataset

Once the dataset has been downloaded (and compressed), pandas can just read the CSV file straight out of the zip file:

In [7]:
prosper_load_data_df = pd.read_csv(prosper_load_data_csv_zip)

### What is the structure of your dataset?

In [8]:
prosper_load_data_df.sample(5)

ListingKey  ListingNumber            ListingCreationDate  \
37997  9ADD35189255014232E18F8         512955  2011-06-24 08:26:21.163000000   
91838  710635296492819282F5167         536431  2011-10-31 09:21:08.467000000   
70631  E58E35427966827697CE2FF         571521  2012-03-23 09:50:24.020000000   
70344  EECA35983938011039F2B18        1115285  2014-01-02 21:35:38.043000000   
79175  E37034573595920626F266C         417512  2009-07-21 20:37:20.303000000   

      CreditGrade  Term            LoanStatus           ClosedDate  \
37997         NaN    60             Completed  2012-08-28 00:00:00   
91838         NaN    36  Past Due (1-15 days)                  NaN   
70631         NaN    36               Current                  NaN   
70344         NaN    36               Current                  NaN   
79175         NaN    36             Completed  2012-05-21 00:00:00   

       BorrowerAPR  BorrowerRate  LenderYield  ...  LP_ServiceFees  \
37997      0.20436        0.1899       0.1799  ...          -95.40   
91838      0.35132        0.3149       0.3049  ...         -124.26   
70631      0.26681        0.2287       0.2187  ...         -216.72   
70344      0.15223        0.1239       0.1139  ...           -3.40   
79175      0.22491        0.2000       0.1900  ...          -25.08   

       LP_CollectionFees  LP_GrossPrincipalLoss  LP_NetPrincipalLoss  \
37997               0.00                    0.0                  0.0   
91838             -25.96                    0.0                  0.0   
70631               0.00                    0.0                  0.0   
70344               0.00                    0.0                  0.0   
79175               0.00                    0.0                  0.0   

      LP_NonPrincipalRecoverypayments  PercentFunded  Recommendations  \
37997                             0.0         0.9113                0   
91838                             0.0         1.0000                0   
70631                             0.0         1.0000                0   
70344                             0.0         1.0000                0   
79175                             0.0         1.0000                0   

      InvestmentFromFriendsCount InvestmentFromFriendsAmount Investors  
37997                          0                         0.0       161  
91838                          0                         0.0       103  
70631                          0                         0.0       225  
70344                          0                         0.0         1  
79175                          0                         0.0        23  

[5 rows x 81 columns]

In [9]:
prosper_load_data_df.shape

(113937, 81)

In [10]:
prosper_load_data_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 113937 entries, 0 to 113936
Data columns (total 81 columns):
 #   Column                               Non-Null Count   Dtype  
---  ------                               --------------   -----  
 0   ListingKey                           113937 non-null  object 
 1   ListingNumber                        113937 non-null  int64  
 2   ListingCreationDate                  113937 non-null  object 
 3   CreditGrade                          28953 non-null   object 
 4   Term                                 113937 non-null  int64  
 5   LoanStatus                           113937 non-null  object 
 6   ClosedDate                           55089 non-null   object 
 7   BorrowerAPR                          113912 non-null  float64
 8   BorrowerRate                         113937 non-null  float64
 9   LenderYield                          113937 non-null  float64
 10  EstimatedEffectiveYield              84853 non-null   float64
 11  EstimatedLoss

Before we delve into the contents, we need to make sure certain basic data hygiene criteria are met. Are there any duplicate rows?

In [13]:
prosper_load_data_df.duplicated().sum()

0

Are there multiple records with the same ``ListingKey`` values?

In [14]:
prosper_load_data_df.ListingKey.duplicated().sum()

871

There are entries with the same ``ListingKey``. Let's explore this records to determine whether we need to drop the duplicates or not.

In [18]:
prosper_load_data_df[prosper_load_data_df.ListingKey.duplicated(keep=False)].sort_values("ListingKey").head(10)

ListingKey  ListingNumber            ListingCreationDate  \
32680  00223594917038064A7C947         998257  2013-11-15 16:58:37.167000000   
32681  00223594917038064A7C947         998257  2013-11-15 16:58:37.167000000   
32964  00473590513960687DD308F         941296  2013-10-07 15:47:36.023000000   
17274  00473590513960687DD308F         941296  2013-10-07 15:47:36.023000000   
7478   0098360461900952056DB93        1190614  2014-03-02 14:21:39.583000000   
33220  0098360461900952056DB93        1190614  2014-03-02 14:21:39.583000000   
27677  01163604029146842E28D9C        1233732  2014-02-25 14:33:39.830000000   
788    01163604029146842E28D9C        1233732  2014-02-25 14:33:39.830000000   
27360  014F35910923350802E1B29         930618  2013-09-26 16:44:24.163000000   
18324  014F35910923350802E1B29         930618  2013-09-26 16:44:24.163000000   

      CreditGrade  Term LoanStatus ClosedDate  BorrowerAPR  BorrowerRate  \
32680         NaN    36    Current        NaN      0.23540        0.1980   
32681         NaN    36    Current        NaN      0.23540        0.1980   
32964         NaN    60    Current        NaN      0.22242        0.1980   
17274         NaN    60    Current        NaN      0.22242        0.1980   
7478          NaN    36    Current        NaN      0.29660        0.2574   
33220         NaN    36    Current        NaN      0.29660        0.2574   
27677         NaN    36    Current        NaN      0.31975        0.2800   
788           NaN    36    Current        NaN      0.31975        0.2800   
27360         NaN    60    Current        NaN      0.19323        0.1695   
18324         NaN    60    Current        NaN      0.19323        0.1695   

       LenderYield  ...  LP_ServiceFees  LP_CollectionFees  \
32680       0.1880  ...          -12.34                0.0   
32681       0.1880  ...          -12.34                0.0   
32964       0.1880  ...          -73.12                0.0   
17274       0.1880  ...          -73.12                0.0   
7478        0.2474  ...            0.00                0.0   
33220       0.2474  ...            0.00                0.0   
27677       0.2700  ...            0.00                0.0   
788         0.2700  ...            0.00                0.0   
27360       0.1595  ...          -27.50                0.0   
18324       0.1595  ...          -27.50                0.0   

       LP_GrossPrincipalLoss  LP_NetPrincipalLoss  \
32680                    0.0                  0.0   
32681                    0.0                  0.0   
32964                    0.0                  0.0   
17274                    0.0                  0.0   
7478                     0.0                  0.0   
33220                    0.0                  0.0   
27677                    0.0                  0.0   
788                      0.0                  0.0   
27360                    0.0                  0.0   
18324                    0.0                  0.0   

      LP_NonPrincipalRecoverypayments  PercentFunded  Recommendations  \
32680                             0.0            1.0                0   
32681                             0.0            1.0                0   
32964                             0.0            1.0                0   
17274                             0.0            1.0                0   
7478                              0.0            1.0                0   
33220                             0.0            1.0                0   
27677                             0.0            1.0                0   
788                               0.0            1.0                0   
27360                             0.0            1.0                0   
18324                             0.0            1.0                0   

      InvestmentFromFriendsCount InvestmentFromFriendsAmount Investors  
32680                          0                         0.0         1  
32681                          0                         0.0         1  
32964       

From these sample, we can assume that there are all duplicated entries which can be safely removed. Notice that we assign the result of dropping the duplicates to a new variable ``prosper_load_data_clean_df`` to keep the original data as reference.

In [19]:
prosper_load_data_clean_df = prosper_load_data_df[~prosper_load_data_df.ListingKey.duplicated()]

In [21]:
prosper_load_data_clean_df.shape

(113066, 81)

The dataset is comprised of 113937 rows (loans) and 81 columns. The purpose of each column is described in the [data dictionary](#section).

### What is/are the main feature(s) of interest in your dataset?

### What features in the dataset do you think will help support your investigation into your feature(s) of interest?

> Your answer here!

## Univariate Exploration

> In this section, investigate distributions of individual variables. If
you see unusual points or outliers, take a deeper look to clean things up
and prepare yourself to look at relationships between variables.


> **Rubric Tip**: The project (Parts I alone) should have at least 15 visualizations distributed over univariate, bivariate, and multivariate plots to explore many relationships in the data set.  Use reasoning to justify the flow of the exploration.



>**Rubric Tip**: Use the "Question-Visualization-Observations" framework  throughout the exploration. This framework involves **asking a question from the data, creating a visualization to find answers, and then recording observations after each visualisation.** 




>**Rubric Tip**: Visualizations should depict the data appropriately so that the plots are easily interpretable. You should choose an appropriate plot type, data encodings, and formatting as needed. The formatting may include setting/adding the title, labels, legend, and comments. Also, do not overplot or incorrectly plot ordinal data.

### Discuss the distribution(s) of your variable(s) of interest. Were there any unusual points? Did you need to perform any transformations?

> Your answer here!

### Of the features you investigated, were there any unusual distributions? Did you perform any operations on the data to tidy, adjust, or change the form of the data? If so, why did you do this?

> Your answer here!

## Bivariate Exploration

> In this section, investigate relationships between pairs of variables in your
data. Make sure the variables that you cover here have been introduced in some
fashion in the previous section (univariate exploration).

### Talk about some of the relationships you observed in this part of the investigation. How did the feature(s) of interest vary with other features in the dataset?

> Your answer here!

### Did you observe any interesting relationships between the other features (not the main feature(s) of interest)?

> Your answer here!

## Multivariate Exploration

> Create plots of three or more variables to investigate your data even
further. Make sure that your investigations are justified, and follow from
your work in the previous sections.

### Talk about some of the relationships you observed in this part of the investigation. Were there features that strengthened each other in terms of looking at your feature(s) of interest?

> Your answer here!

### Were there any interesting or surprising interactions between features?

> Your answer here!

## Conclusions
>You can write a summary of the main findings and reflect on the steps taken during the data exploration.



> Remove all Tips mentioned above, before you convert this notebook to PDF/HTML


> At the end of your report, make sure that you export the notebook as an
html file from the `File > Download as... > HTML or PDF` menu. Make sure you keep
track of where the exported file goes, so you can put it in the same folder
as this notebook for project submission. Also, make sure you remove all of
the quote-formatted guide notes like this one before you finish your report!

